### Sampler를 이용한 실험 결과 (전체)
- Test f1 score 기준 내림차순 정렬

In [1]:
import pandas as pd
pd.set_option('display.max_colwidth', None)

df = pd.read_csv("../experimental_records.csv")
df = df[df['model_name'].apply(lambda x: True if x.find('sampling') + 1 else False)]
df.sort_values(by='test_f1_score', ascending=False)

date  \
21  2024-02-12/00:52:24   
22  2024-02-12/01:14:02   
24  2024-02-12/11:45:10   
23  2024-02-12/01:55:22   
14  2024-02-08/16:18:57   
16  2024-02-11/16:29:16   
13  2024-02-07/18:27:07   
15  2024-02-11/16:22:37   
20  2024-02-12/00:02:37   
19  2024-02-11/18:20:36   
6   2024-02-06/00:00:20   
12  2024-02-07/18:22:53   
7   2024-02-06/11:45:07   
11  2024-02-06/15:20:02   
17  2024-02-11/18:13:58   
8   2024-02-06/14:51:02   
18  2024-02-11/18:17:33   
5   2024-02-05/22:53:27   
9   2024-02-06/15:02:31   
10  2024-02-06/15:12:31   
4   2024-02-05/17:56:50   

                                                       model_name  \
21                  ensemble_decisiontree_w_undersampling_and_ccp   
22           ensemble_decisiontree_w_undersampling_ccp_delFeature   
24             ensemble_decisiontree_w_undersampling_bestccpAlpha   
23  ensemble_decisiontree_w_undersampling_bestccpAlpha_delFeature   
14           ensemble_decisiontree_w_undersampling_ccp_delFeature   
16          ensemble_decisiontree_w_undersampling_ccp_delFeatures   
13                  ensemble_decisiontree_w_undersampling_and_ccp   
15          ensemble_decisiontree_w_undersampling_ccp_delFeatures   
20                          ensemble_decisiontree_w_undersampling   
19                          ensemble_decisiontree_w_undersampling   
6                           ensemble_decisiontree_w_undersampling   
12                 ensemble_decisiontree_w_undersampling_and_seed   
7                           ensemble_decisiontree_w_undersampling   
11       ensemble_decisiontree_w_undersampling_and_delete_feature   
17                          ensemble_decisiontree_w_undersampling   
8                    decision_tree_w_undersampling_delete_feature   
18                          ensemble_decisiontree_w_undersampling   
5                                   decision_tree_w_undersampling   
9                   decision_tree_w_undersampling_delete_features   
10            decision_tree_w_undersampling_feature_normalization   
4                                    decision_tree_w_oversampling   

    test_f1_score  \
21       0.712175   
22       0.703117   
24       0.696502   
23       0.696502   
14       0.695737   
16       0.693580   
13       0.691334   
15       0.691334   
20       0.669623   
19       0.665550   
6        0.652855   
12       0.651697   
7        0.649734   
11       0.638188   
17       0.623391   
8        0.610570   
18       0.592551   
5        0.586722   
9        0.581432   
10       0.554104   
4        0.391892   

                                                                                                                                                                                                                      description  
21                                                                                           30개 decision tree 사용, decision tree마다 seed 세팅/seed33~62/ccp_alpha 0.00045 사용/mean validation f1 score 0.8960/hard voting/결측치 0으로 처리/  
22                                                           30개 decision tree 사용, decision tree마다 seed 세팅/seed33~62/ccp_alpha 0.0004 사용/"customer_country.1" feature 삭제/mean validation f1 score 0.89571/hard voting/결측치 0으로 처리/  
24                                                                                    30개 decision tree 사용, decision tree마다 seed 세팅/seed33~62/모델마다 best ccp alpha 찾아서 적용/mean validation f1 score 0.90029/hard voting/결측치 0으로 처리/  
23                                                    30개 decision tree 사용, decision tree마다 seed 세팅/seed33~62/모델마다 best ccp alpha 찾아서 적용/"customer_country.1" feature 삭제/mean validation f1 score 0.90029/hard voting/결측치 0으로 처리/  
14                                                           10개 decision tree 사용, decision tree마다 seed 세팅/seed33~52/ccp_alpha 0.0004 사용/"customer_country.1" feature 삭제/mean validation f1 score 0.89659/hard voting/결측치 0으로 처리/  
16  10개 decision tree 사용, decision tree마다 seed 세팅/seed33~52/ccp_alpha 0.00

---
---

**<span style='color:blue'>[0.391892]</span> 실험 01: Positive sample oversampling**

**Motivation**
- Data imbalance 문제를 positive sample에 대해 oversampling 함으로써 해결해보고자 시도

**Test f1 score**
- 0.391892 (-0.083564) | Best: 0.475456

**Analysis**
- Negative sample이 positive sample보다 10배 이상 많은 상황에서 positive sample을 oversampling했기 때문에 중복되는 positive sample이 너무 많아졌음.
- 중복되는 positive sample들을 올바르게 분류하도록 학습하다보니 분류 기준이 이상해지게 되었고 (overfitting), 그로 인해 test f1 score가 크게 하락한 것으로 분석하였음.

**Future works**
- Undersampling
---
---

**<span style='color:red'>[0.586722]</span> 실험 02: Negative sample undersampling**

**Motivation**
- 데이터 수가 훨씬 많은 negative sample을 undersampling하여 positive sample과 비율을 맞춰준다면, training data에 overfitting 되는 현상을 완화할 수 있을 것으로 기대

**Test f1 score**
- 0.586722 (+0.111266) | Best: 0.475456

**Analysis**
- Decision tree의 특성 상, 특정 class의 samples이 많은 경우 해당 samples를 잘 분류하도록 만들어지기 때문에 bias가 생길 수 있음.
- Sampling을 통해서 적어도 negative / positive class 간의 imbalance는 해소해주었기 때문에, test f1 score가 큰 폭으로 향상될 수 있었다고 분석하였음.

**Future works**
- 버려지는 negative samples를 활용할 수 있는 방법을 찾아볼 것
---

**<span style='color:red'>[0.669623]</span> 실험 03: Ensemble with undersampling**

**Motivation**
- Positive samples의 수에 맞게 negative samples를 undersampling하게 되면서 버려지는 negative samples이 너무 많았음.
- 서로 다른 seed 값을 사용해 sampling 함으로써 서로 다른 data subset을 만들고, 각각의 data subset을 학습한 모델들을 ensemble 한다면, 특정 subset 에만 overfitting 되는 문제를 완화할 수 있을 것으로 기대.

**Test f1 score**
- Previous best test f1 score: 0.586722
- 사용한 voting : hard voting (과반수를 초과하는 모델이 1이라고 예측 시 1, 이하는 0)
- Ensemble한 모델의 개수에 따른 test f1 score table

|num_models|test f1 score|
|:-:|:-:|
| 5 | 0.623391 (+0.036669) |
| 10 | 0.592551 (+0.005829) |
| 20 | 0.665550 (+0.078828) |
| 30 | 0.669623 (+0.082901) |

**Analysis**
- Negative samples은 subset마다 다르고 positive samples은 모든 subset이 동일함.
- Subset마다 다른 negative samples로 인해 각각의 classification tree는 약간씩 다른 분류 기준을 가지게 되고, 예측 결과에 있어 다양성을 가질 수 있을 것이라 생각함.
- 모든 classification tree가 같은 positive samples을 학습했기 때문에, 이와 비슷한 test data가 입력되는 경우 아주 높은 확률로 positive samples로 예측할 것.
- 반면에 모든 classification tree가 약간씩 다른 negative samples을 학습했기 때문에, 임의의 test data가 입력되는 경우 voting을 통해서 negative samples을 판단하게 될 것.
- 본 실험에서 test f1 score가 크게 높아질 수 있었던 주된 이유는 negative samples의 분류 능력이 높아졌기 때문으로 분석함.
---
- 모델 개수에 따른 성능 차이를 비교분석했을 때, 30 개의 모델을 ensemble 했을 때가 가장 높은 성능을 기록
- 10개일 때 성능이 떨어진 것은 seed 의 영향으로 보임

**Future works**
- positive samples을 증강할 수 있는 방법
- feature preprocessing
- pruning
---
---

**<span style='color:red'>[0.610570]</span> 실험 04: Decision tree with undersampling & feature preprocessing**

**Motivation**
- Sampling 없이, feature preprocessing 후 classification tree를 모델링했을 때 모두 성능이 하락하였으나, 성능 하락의 원인이 불분명하다고 판단하여 재실험을 진행하고자 하였음.
- 총 3가지의 시나리오를 수행

**Test f1 score**
- 비교군: 실험 B (0.586722)

|scenario|test f1 score|
|:-:|:-:|
|`customer_country.1` feature 삭제|0.610570 (+0.023848)|
|`customer_country.1` feature 삭제 + 결측치 비율 80% 이상인 features 삭제|0.581432 (-0.00529)|
|`customer_country.1` feature 삭제 + `customer_country` feature 정규화|0.554104 (-0.032618)|

**Analysis**
- scenario 01: `customer_country` feature 삭제
    - `customer_country` 와 `customer_country.1` feature 간의 correlation이 1에 가까울 정도로 유사하다는 점을 EDA를 통해 확인하였고, 둘 중 하나를 삭제하여 불필요한 정보를 학습하지 못하도록 방지하고자 하였음.
    - 실험 B 대비 큰 폭으로 향상
- scenario 02: `customer_country.1` feature 삭제 + 결측치 비율 80% 이상인 features 삭제
    - 현재 결측치는 모두 0으로 채우고 있기 때문에, 결측치 비율이 80% 이상이라면 학습이 무의미하다고 생각하여 삭제하고자 하였음.
    - 삭제한 features : `['id_strategic_ver', 'it_strategic_ver', 'idit_strategic_ver',
       'product_subcategory', 'product_modelname', 'business_subarea','customer_country.1']`
    - 한 번에 너무 많은 features를 삭제하면서 성능 하락의 원인이 불분명 -> 추가 실험을 진행해야할 것으로 보임
    - 확실한 것은 결측치 비율이 높다고 해서 feature를 삭제하는 것은 오히려 성능 하락의 원인이 된다는 것
- scenario 03: `customer_country.1` feature 삭제 + `customer_country` feature 정규화
    - `customer_country` feature는 고객이 기입한 국적 정보인데, 나라 뿐만 아니라 도시 정보도 함께 기입되면서 종류의 수가 너무 많았음. 세부 정보들이 불필요한 정보들이라고 판단하여 도시 정보만 남기는 방식으로 정규화를 시도, 성능을 확인하였으나 오히려 큰 폭으로 성능이 떨어짐.
    - 상세한 국적 정보가 target 예측에 오히려 도움이 되었다는 것을 확인.

**Future works**
- 결측치 비율이 높은 features에 대한 ablation study
- 결측값을 어떤 값으로 채울 것인가?
- `customer_country` feature를 오히려 세분화하여 feature의 양을 늘린다면?
---
---

**<span style='color:orange'>[0.712175, best]</span> 실험 05: Cost-Complexity Pruning**

**Motivation**
- 실험 03을 통해 서로 다른 subset을 학습한 모델을 ensemble 했을 때 generalization 성능이 향상된 것을 확인하였음.
- 다만 각각의 모델이 자기가 학습한 subset에 대해 overfitting 되는 문제는 여전히 남아있었음.
- Cost-complexity pruning을 통해 overfitting 문제를 완화한 뒤 ensemble 한다면 generalization 성능을 더 높일 수 있을 것이라 판단하여 실험을 진행
- 총 30개의 모델을 ensemble 하였음

**Test f1 score**
- `customer_country.1` feature를 삭제한 상태에서, 두 가지 시나리오를 실험.
- Prev best test f1 score: 0.669623

|scenario|mean validation f1 score|test f1 score|
|:-:|:-:|:-:|
|임의의 subset 하나에 대한 best ccp_alpha (0.00045) 를 모든 모델에 적용|0.8960|0.712175 (+0.042552)|
|subset마다 best ccp_alpha를 구해 학습|0.90029|0.696502 (+0.026879)|

**Analysis**
- CCP를 통해 분류 결과가 극도로 세분화되는 문제를 완화 (training data만 잘 분류할 수 있도록 모델링되는 문제를 완화)
- Overfitting 문제가 완화된 모델들을 ensemble 하면서 generalization 성능이 극대화된 것으로 분석
- 조금 의아했던 것은 subset별 best ccp_alpha 값을 구해서 적용했을 때 성능이 조금 더 높아질 것이라고 예상했던 것과 반대의 결과가 나왔다는 것
- 러프하게 ccp_alpha 값을 적용한 것과 subset별 best ccp_alpha 값을 적용한 것을 비교했을 때, 전자에 비해 후자가 subset에 좀 더 fitting되면서 general한 예측을 하지 못하게 된 것이라고 분석하였음.

**Future works**
- feature preprocessing과 결합
---
---

**<span style='color:blue'>[0.703117]</span> 실험 06: Cost-Complexity Pruning + `customer_country.1` feature 삭제**

**Motivation**
- 실험 04-01 을 통해 `customer_country.1` feature를 삭제했을 때 test f1 score 가 높아지는 것을 확인
- 따라서 이전의 best 모델이었던 실험 05 세팅에 04-01 세팅을 합친다면 test f1 score 를 더 높일 수 있을 것이라 판단하고 실험을 진행
- 총 30개의 모델을 ensemble 하였음.

**Test f1 score**
- `customer_country.1` feature를 삭제한 상태에서, 두 가지 시나리오를 실험.

|scenario|mean validation f1 score|test f1 score|
|:-:|:-:|:-:|
|subset 하나에 대한 best ccp_alpha를 모든 모델에 적용|0.89571|0.703117 (-0.009058)|
|subset마다 best ccp_alpha를 구해 학습|0.90029|0.696502 (-0.015673)|

**Analysis**
- Feature 간에 correlation이 1에 가깝다는 것은 분포가 거의 동일하다는 얘기임.
- 이 말은 target을 예측할 때 둘 중 하나는 없어도 무방하다는 의미.
- 허나 `customer_country.1` feature를 삭제했을 때 성능이 하락하였음.
- 여러 subset을 만들고 이를 학습하는 과정에서, 특정 subset의 경우 `customer_country.1` feature의 중요도가 높았을 수도 있음.
- `customer_country.1` feature의 중요도가 높았던 subset 들에서 feature를 제거하게 되면서, 해당 subset을 학습한 모델의 분류 성능은 떨어지게 되고, 이러한 모델들이 모이게 되면서 최종적으로 성능이 떨어진 게 아닌가 싶음.
---
---

### 정리

`+`
- Negative samples undersampling
- Ensemble
- Cost-Complexity Pruning
---
`-`
- Positive samples oversampling
- Feature preprocessing
---
---

### Future works

- 모델 (e.g., svm, gbm, AdaBoost, lightGBM, XGBoost, CatBoost) 변경
- 새로운 기법 탐색 (e.g., Logistic Regression, Out-Of-Distribution Detection)
- 결측값 처리
- feature engineering
- data augmentation